docs(weekly): add this week in databend 113 (#13086)
datafuselabs/databend: website/blog/2023-10-01-databend-weekly-113.md

**NoPrefetchPoliy** 

No prefetch stage at all. Databend reads, deserializes, and outputs data blocks you need directly.

**PredicateAndTopkPolicy**

Databend prefetches columns needed by prewhere and topk at the prefetch stage. It deserializes them into a DataBlock and evaluates them into RowSelection. It then slices the DataBlock by batch size and stores the resulting VecDeque in memory.

Databend reads the remaining columns specified by RowSelection at the final stage, and it outputs DataBlocks in batches. Then, it merges the prefetched blocks and projects the resulting blocks according to the output_schema.

**TopkOnlyPolicy**

It's similar to the `PredicateAndTopkPolicy`, but Databend only evaluates the topk at the prefech stage.

If you are interested in learning more, please check out the resources below:

- [PR #13020 | refactor: introduce ReadPolicy to parquet reader](https://github.com/datafuselabs/databend/pull/13020)

## Highlights

We have also made these improvements to Databend that we hope you will find helpful:

- Added spill info to query log.

> - Added support for compression when using `COPY INTO`.

Reviewer: ```suggestion
- Added support for unloading data into a compressed file with COPY INTO.
```

  - name: zhang2014
  - name: ZhiHanZ
  - name: zhyass
authors:
  - name: PsiACE
    url: https://github.com/psiace
    image_url: https://github.com/psiace.png
---

[Databend](https://github.com/datafuselabs/databend) is a modern cloud data warehouse, serving your massive-scale analytics needs at low cost and complexity. Open source alternative to Snowflake. Also available in the cloud: <https://app.databend.com> .

## What's On In Databend

Stay connected with the latest news about Databend.

### Loading to Table with Extra Columns

By default, COPY INTO loads data into a table by matching the order of fields in the file to the corresponding columns in the table. It's essential to ensure that the data aligns correctly between the file and the table.

![load extra](https://databend.rs/assets/images/load-extra-0b76668838a386aace5463d65547a40f.png)

If your table has more columns than the file, you can specify the columns into which you want to load data.

When working with CSV format, if your table has more columns than the file and the additional columns are at the end of the table, you can load data using the `FILE_FORMAT` option `ERROR_ON_COLUMN_COUNT_MISMATCH`.

> If you are interested in learning more, please check out the resources listed below.

Reviewer: ```suggestion
If you are interested in learning more, please check out the resources below:
```

- [Docs | Example 5: Loading to Table with Extra Columns](https://databend.rs/doc/sql-commands/dml/dml-copy-into-table#example-5-loading-to-table-with-extra-columns)

## Code Corner

Discover some fascinating code snippets or projects that showcase our work or learning journey.

### Introducing Read Policies to Parquet Reader

> There is a drawback of using `arrow-rs` APIs: when we try to prefetching data for prewhere and topk push downs, we can't reuse the deserialized blocks.

Reviewer: ```suggestion
There is a drawback of using `arrow-rs` APIs: When we prefetch data for prewhere and topk push downs, we can't reuse the deserialized blocks.
```
```suggestion
There is a drawback of using `arrow-rs` APIs: when we try to prefetching data for prewhere and topk push downs, we can't reuse the deserialized blocks.
```

> There is a drawback of using `arrow-rs` APIs: when we try to prefetching data for prewhere and topk push downs, we can't reuse the deserialized blocks.

Reviewer: ```suggestion
There is a drawback of using `arrow-rs` APIs: When we try to prefetch data for prewhere and topk push downs, we can't reuse the deserialized blocks.
```

In order to improve the logic of row group reading and reuse the prefetched data blocks at the final output stage, we have done a lot of refactoring and introduced some read policies.

**NoPrefetchPoliy** 

> There is no prefetch stage, we can read, deserialize and output data block we need directly.

Reviewer: ```suggestion
No prefetch stage at all. Read, deserialize, and output data blocks you need directly.
```

**PredicateAndTopkPolicy**

> At prefetch stage, prefetch columns needed by prewhere and topk. Deserialize them into `DataBlock` and evaluate it into `RowSelection`. Slice the `DataBlock` by batch size and store the `VecDeque` in memory.

Reviewer: ```suggestion
Databend prefetches columns needed by prewhere and topk at the prefetch stage. It deserializes them into a DataBlock and evaluates them into RowSelection. It then slices the DataBlock by batch size and stores the resulting VecDeque in memory.
```

> At final stage, read remain columns by `RowSelection` and output `DataBlock`s by batch size, and merge the prefetch block and project the block by `output_schema`.

Reviewer: ```suggestion
Databend reads the remaining columns specified by RowSelection at the final stage, and it outputs DataBlocks in batches. Then, it merges the prefetched blocks and projects the resulting blocks according to the output_schema.
```

**TopkOnlyPolicy**

> It's similar to `PredicateAndTopkPolicy`, but we only evaluate topk at prefech stage.

Reviewer: ```suggestion
It's similar to the `PredicateAndTopkPolicy`, but Databend only evaluates topk at the prefech stage.
```

> If you are interested in learning more, please check out the resources listed below:

Reviewer: ```suggestion
If you are interested in learning more, please check out the resources below:
```

- [PR #13020 | refactor: introduce ReadPolicy to parquet reader](https://github.com/datafuselabs/databend/pull/13020)

## Highlights

We have also made these improvements to Databend that we hope you will find helpful:

- Added spill info to query log.

> - Added support for compression when using `COPY INTO`.

Reviewer: Could you pls send me the PR link? Thanks.

Author: https://github.com/datafuselabs/databend/pull/13067

> - Added `GET /v1/background/:tenant/background_tasks` http api to query background tasks.

Reviewer: ```suggestion
- Introduced the `GET /v1/background/:tenant/background_tasks` HTTP API for querying background tasks.
```

- Read [Example 4: Filtering Files with Pattern](https://databend.rs/doc/sql-commands/dml/dml-copy-into-table#example-4-filtering-files-with-pattern) to understand how to use Pattern to filter files.

## What's Up Next

We're always open to cutting-edge technologies and innovative ideas. You're more than welcome to join the community and bring them to Databend.

> ### Fixing the issues detected by SQLsmith

Reviewer: ```suggestion
### Fixing issues detected by SQLsmith
```

> So far SQLsmith has found about 40 panics or other bugs in Databend, in the past month since it’s been running. Databend Labs has been working on fixing these issues to improve stability, even in edge cases.

Reviewer: ```suggestion
In the last month, SQLsmith has discovered around 40 bugs in Databend. Databend Labs is actively working to fix these issues and improve system stability, even in uncommon scenarios. Your involvement in this effort, which may include tasks like type conversion or handling special values, is encouraged and can be facilitated by referring to past fixes.
```

> We also hope that you can participate in this work, which may involve type conversion or special values. By referring to previous fixes, it can be easily fixed.

Reviewer: ```suggestion

```
